federation/pkg/federation-controller/namespace/namespace_controller.go

```diff
@@ -147,24 +148,118 @@ func NewNamespaceController(client federationclientset.Interface) *NamespaceCont
 
 	// Federated updeater along with Create/Update/Delete operations.
 	nc.federatedUpdater = util.NewFederatedUpdater(nc.namespaceFederatedInformer,
-		func(client kubeclientset.Interface, obj pkg_runtime.Object) error {
+		func(client kubeclientset.Interface, obj runtime.Object) error {
 			namespace := obj.(*api_v1.Namespace)
 			_, err := client.Core().Namespaces().Create(namespace)
 			return err
 		},
-		func(client kubeclientset.Interface, obj pkg_runtime.Object) error {
+		func(client kubeclientset.Interface, obj runtime.Object) error {
 			namespace := obj.(*api_v1.Namespace)
 			_, err := client.Core().Namespaces().Update(namespace)
 			return err
 		},
-		func(client kubeclientset.Interface, obj pkg_runtime.Object) error {
+		func(client kubeclientset.Interface, obj runtime.Object) error {
 			namespace := obj.(*api_v1.Namespace)
 			err := client.Core().Namespaces().Delete(namespace.Name, &api_v1.DeleteOptions{})
+			// IsNotFound error is fine since that means the object is deleted already.
+			if errors.IsNotFound(err) {
+				return nil
+			}
 			return err
 		})
+
+	nc.deletionHelper = deletionhelper.NewDeletionHelper(
+		nc.hasFinalizerFunc,
+		nc.removeFinalizerFunc,
+		nc.addFinalizerFunc,
+		// objNameFunc
+		func(obj runtime.Object) string {
+			namespace := obj.(*api_v1.Namespace)
+			return namespace.Name
+		},
+		nc.updateTimeout,
+		nc.eventRecorder,
+		nc.namespaceFederatedInformer,
+		nc.federatedUpdater,
+	)
 	return nc
 }
 
+// Returns true if the given object has the given finalizer in its ObjectMeta.
+func (nc *NamespaceController) hasFinalizerFunc(obj runtime.Object, finalizer string) bool {
+	namespace := obj.(*api_v1.Namespace)
+	for i := range namespace.ObjectMeta.Finalizers {
+		if string(namespace.ObjectMeta.Finalizers[i]) == finalizer {
+			return true
+		}
+	}
+	return false
+}
+
+// Removes the finalizer from the given objects ObjectMeta.
+// Assumes that the given object is a namespace.
+func (nc *NamespaceController) removeFinalizerFunc(obj runtime.Object, finalizer string) (runtime.Object, error) {
+	namespace := obj.(*api_v1.Namespace)
+	newFinalizers := []string{}
+	hasFinalizer := false
+	for i := range namespace.ObjectMeta.Finalizers {
+		if string(namespace.ObjectMeta.Finalizers[i]) != finalizer {
+			newFinalizers = append(newFinalizers, namespace.ObjectMeta.Finalizers[i])
+		} else {
+			hasFinalizer = true
+		}
+	}
+	if !hasFinalizer {
+		// Nothing to do.
+		return obj, nil
+	}
+	namespace.ObjectMeta.Finalizers = newFinalizers
+	namespace, err := nc.federatedApiClient.Core().Namespaces().Finalize(namespace)
```

Comment:
Why not just call Update? Calling `Finalize` looks like it's delete the namespace.spec.finalizer